xla/service/gpu/gpu_all_gather_optimizer.cc

```diff
@@ -0,0 +1,114 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/gpu/gpu_all_gather_optimizer.h"
+
+#include "xla/hlo/ir/hlo_casting_utils.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_instructions.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/service/collective_ops_utils.h"
+#include "xla/hlo/utils/hlo_query.h"
+#include "xla/service/reduce_scatter_utils.h"
+
+namespace xla {
+namespace gpu {
+
+StatusOr<bool> AllGatherOptimizer::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  const HloModuleConfig& config = module->config();
+  int64_t next_channel_id = hlo_query::NextChannelId(*module);
+
+  bool changed = false;
+  for (HloComputation* computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction* instruction :
+         computation->MakeInstructionPostOrder()) {
+      if (!HloOpcodeIsBinaryCommutative(instruction->opcode())) {
+        continue;
+      }
+
+      HloInstruction* left_op = instruction->mutable_operand(0);
+      HloInstruction* right_op = instruction->mutable_operand(1);
+
+      // we need to traverse the right branch backwards in search of
+      // an all-gather collective
+      auto find_all_gather = [&](HloInstruction* base_op) {
```

Comment:
Is it actually correct if you apply this transformation if the two all_gathers are not directly the operands of the binary op?

Reply:
couldn't prove it for every car though it passed all the use cases I had. I removed the traversal.